optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,281 @@
+from collections import defaultdict
+import itertools
+from typing import Any
+from typing import DefaultDict
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Tuple
+
+import numpy as np
+
+import optuna
+from optuna._experimental import experimental
+from optuna.distributions import BaseDistribution
+from optuna import multi_objective
+from optuna.multi_objective.samplers import BaseMultiObjectiveSampler
+
+
+# Define key names of `Trial.system_attrs`.
+_GENERATION_KEY = "multi_objective:nsga2:generation"
+_PARENTS_KEY = "multi_objective:nsga2:parents"
+
+
+@experimental("1.5.0")
+class NSGAIIMultiObjectiveSampler(BaseMultiObjectiveSampler):
+    """Multi-objective sampler using the NSGA-II algorithm.
+
+    NSGA-II stands for "Nondominated Sorting Genetic Algorithm II",
+    which is a well known, fast and elitist multi-objective genetic algorithm.
+
+    For further information about NSGA-II, please refer to the following paper:
+
+    - `A fast and elitist multiobjective genetic algorithm: NSGA-II
+      <https://ieeexplore.ieee.org/document/996017>`_
+
+    Args:
+        seed:
+            Seed for random number generator.
+
+        population_size:
+            Number of individuals (trials) in a generation.
+
+        mutation_prob:
+            Probability of mutating each parameter when creating a new individual.
+            If :obj:`None` is specified, the value ``1.0 / len(parent_trial.params)`` is used
+            where ``parent_trial`` is the first parent trial of the target individual.
```

Comment:
How about abstracting it a bit, now that we also try the "second" parent?
```suggestion
            where ``parent_trial`` is the parent trial of the target individual.
```